### PR TITLE
Finish updating spago CI to 0.15

### PR DESCRIPTION
**Pull request by JordanMartinez**
### Description of the change

Updates CI now that we have a working `0.15.0` package set.

### Checklist:

- [x] Added the change to the "Unreleased" section of the changelog
- [ ] Added some example of the new feature to the `README`
- [ ] Added a test for the contribution (if applicable)

**P.S.**: the above checks are not compulsory to get a change merged, so you may skip them. However, taking care of them will result in less work for the maintainers and will be much appreciated 😊

### CODE REVIEW

**Comment by JordanMartinez** [comment]
CI fails currently because `simple-json` hasn't been updated to `0.15.0`. We could workaround that by using some other library for the test or just wait until it gets updated.

**Comment by JordanMartinez** [comment]
Do I need to update the `appveyor.yml` build, too?

**Comment by f-f** [comment]
> CI fails currently because simple-json hasn't been updated to 0.15.0. We could workaround that by using some other library for the test or just wait until it gets updated.

Yep, let's use some library from core instead

> Do I need to update the appveyor.yml build, too?

Yes, that will need to use 0.15 as well

**Comment by JordanMartinez** [comment]
I think I've fixed all CI errors aside from one. The errors were caused by one of three things:
- a different order of dependencies
- using Node 8 rather than the minimum Node 12 (though I opted for 14 since 12 is EOL in a few months)
- not installing esbuild

As for the last error, I think this is a bug in spago:
```
  test\SpagoSpec.hs:86:7: 
  1) Spago, spago init, Spago should import configs from Bower
       expected: "{-\nWelcome to a Spago project!\nYou can edit this file as you like.\n\nNeed help? See the following resources:\n- Spago documentation: https://github.com/purescript/spago\n- Dhall language tour: https://docs.dhall-lang.org/tutorials/Language-Tour.html\n\nWhen creating a new Spago project, you can use\n`spago init --no-comments` or `spago init -C`\nto generate this file without the comments in this block.\n-}\n{ name = \"either\"\n, dependencies =\n  [ \"assert\"\n  , \"console\"\n  , \"control\"\n  , \"effect\"\n  , \"invariant\"\n  , \"maybe\"\n  , \"prelude\"\n  ]\n, packages = ./packages.dhall\n, sources = [ \"src/**/*.purs\", \"test/**/*.purs\" ]\n}\n"
        but got: "{-\nWelcome to a Spago project!\nYou can edit this file as you like.\n\nNeed help? See the following resources:\n- Spago documentation: https://github.com/purescript/spago\n- Dhall language tour: https://docs.dhall-lang.org/tutorials/Language-Tour.html\n\nWhen creating a new Spago project, you can use\n`spago init --no-comments` or `spago init -C`\nto generate this file without the comments in this block.\n-}\n{ name = \"either\"\n, dependencies = [ \"console\", \"effect\", \"prelude\" ]\n, packages = ./packages.dhall\n, sources = [ \"src/**/*.purs\", \"test/**/*.purs\" ]\n}\n"
  To rerun use: --match "/Spago/spago init/Spago should import configs from Bower/"
```
Running the same commands as the test, I see why the test above fails. `spago` doesn't attempt to upgrade the package set before migrating from `bower.json`. In actuality, the error is even more than that. `spago`'s `templates/packages.dhall` file is the `0.14.0` package set, not the `0.15.0` package set. Thus, `spago`:
1. uses the template `packages.dhall` (0.14.x) file to determine "valid" ranges of dependencies
2. attempts to migrate from `bower.json` deps and fails because the deps' versions in `bower.json` are greater than the versions in the `0.14.x` package set.
3. doesn't install any deps from `bower.json`
4. Updates the package set from `0.14.x` to `0.15.x`.

I'm wondering if step 4 should happen before step 1. At the very least, we need two versions of `templates/packages.dhall` and need to use the appropriate one depending on which version of the compiler one is on. Or we could just drop support for `0.14.x` in the next breaking spago change, but that would be pretty annoying during this phase of transition.
```console
$ spago init
[info] Initializing a sample project or migrating an existing one..
[info] Found a "bower.json" file, migrating to a new Spago config..
[warn] 

Spago encountered some errors while trying to migrate your Bower config.
A Spago config has been generated but it's recommended that you apply the suggestions here

These packages are in the set, but did not match the Bower range. You can try to install them with `spago install some-package-name`
* "assert" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "console" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "control" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "effect" has version v3.0.0, but range is >=4.0.0 <5.0.0
* "invariant" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "maybe" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "prelude" has version v5.0.0, but range is >=6.0.0 <7.0.0


[info] Updating package-set tag to "psc-0.15.0-20220502"
Fetching the new one and generating hashes.. (this might take some time)
[info] Generating new hashes for the package set file so it will be cached.. (this might take some time)
[info] Found existing directory "src", skipping copy of sample sources
[info] Found existing directory "test", skipping copy of sample sources
[info] Found existing file ".gitignore", not overwriting it
[info] Found existing file ".purs-repl", not overwriting it
[info] Set up a local Spago project.
[info] Try running `spago build`
$ cat spago.dhall 
{-
Welcome to a Spago project!
You can edit this file as you like.

Need help? See the following resources:
- Spago documentation: https://github.com/purescript/spago
- Dhall language tour: https://docs.dhall-lang.org/tutorials/Language-Tour.html

When creating a new Spago project, you can use
`spago init --no-comments` or `spago init -C`
to generate this file without the comments in this block.
-}
{ name = "either"
, dependencies = [ "console", "effect", "prelude" ]
, packages = ./packages.dhall
, sources = [ "src/**/*.purs", "test/**/*.purs" ]
}
$ rm spago.dhall 
$ spago init
[info] Initializing a sample project or migrating an existing one..
[warn] Found a "packages.dhall" file, skipping copy. Run `spago init --force` if you wish to overwrite it.
[info] Found a "bower.json" file, migrating to a new Spago config..
[info] All Bower dependencies are in the set! 🎉
[info] You can now safely delete your "bower.json"
[info] Found existing directory "src", skipping copy of sample sources
[info] Found existing directory "test", skipping copy of sample sources
[info] Found existing file ".gitignore", not overwriting it
[info] Found existing file ".purs-repl", not overwriting it
[info] Set up a local Spago project.
[info] Try running `spago build`
$ cat spago.dhall 
{-
Welcome to a Spago project!
You can edit this file as you like.

Need help? See the following resources:
- Spago documentation: https://github.com/purescript/spago
- Dhall language tour: https://docs.dhall-lang.org/tutorials/Language-Tour.html

When creating a new Spago project, you can use
`spago init --no-comments` or `spago init -C`
to generate this file without the comments in this block.
-}
{ name = "either"
, dependencies =
  [ "assert", "console", "control", "effect", "invariant", "maybe", "prelude" ]
, packages = ./packages.dhall
, sources = [ "src/**/*.purs", "test/**/*.purs" ]
}
```

**Comment by f-f** [comment]
@JordanMartinez thanks for digging into this! I agree with you that this is a bug, and that the package set upgrade should happen as the first step.
However, I think we don't need to have multiple versions of the template, and we can just upgrade the set as the first step of the import?

**Comment by JordanMartinez** [comment]
> However, I think we don't need to have multiple versions of the template, and we can just upgrade the set as the first step of the import?

I think that will work. The failing test passed locally when I was using PS `0.15.0`. I didn't run any of the other tests.

**Comment by JordanMartinez** [comment]
Huh... any idea why the Appveyor build fails but the GH Windows one doesn't?

**Comment by f-f** [comment]
```

file:///C:/Users/appveyor/AppData/Local/Temp/spago-script-tmp-d04a0b19d4bf9e250e87be4c48982bb282f0822cd43a770abfe170b0e9398af4/.spago/run.js:1\nrequire('C:/Users/appveyor/AppData/Local/Temp/spago-script-tmp-d04a0b19d4bf9e250e87be4c48982bb282f0822cd43a770abfe170b0e9398af4/output/Main').main()\n^\n\nReferenceError: require is not defined in ES module scope, you can use import instead\nThis file is being treated as an ES module because it has a '.js' file extension and 'C:\\Users\\appveyor\\AppData\\Local\\Temp\\spago-script-tmp-d04a0b19d4bf9e250e87be4c48982bb282f0822cd43a770abfe170b0e9398af4\\.spago\\package.json' contains \"type\": \"module\". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.\n    at file:///C:/Users/appveyor/AppData/Local/Temp/spago-script-tmp-d04a0b19d4bf9e250e87be4c48982bb282f0822cd43a770abfe170b0e9398af4/.spago/run.js:1:1\n    at ModuleJob.run (internal/modules/esm/module_job.js:183:25)\n    at async Loader.import (internal/modules/esm/loader.js:178:24)\n    at async Object.loadESM (internal/process/esm_loader.js:68:5)\n    at async handleMainPromise (internal/modules/run_main.js:59:12)\n[error] Script failed to run; exit code: 1\n")
```

Could this be related to the Node version being used?

**Comment by JordanMartinez** [comment]
> Could this be related to the Node version being used?

I don't think so. Both GH and Appveyor are using Node 14.

**Comment by JordanMartinez** [comment]
It looks like the test fails on `0.14`. So, perhaps `spago` is creating the `package.json` file on `0.14` when it shouldn't be? It might help to print which version of `purs` spago detects to see if `0.14` is being detected as `0.14`.

**Comment by JordanMartinez** [comment]
I'm still not sure about the `bower deps` one, but the `spago script` fails because a `packages.json` file exists in the `.spago` dir. Looking at the debug output, it's clear that we are not creating a `packages.json` file in that directory. That means the `package.json` file already existed there before we ran the test.

Sure enough, the directory used for `spago script` is only "semi-temporary": [`<system-temp-dir>/spago-script-tmp-<stableHash>`](https://github.com/purescript/spago/blob/master/src/Spago/Build.hs#L289). Looking at the code, we see that the 'stable hash' will be the same in-between `0.15.0` test runs and `0.14.0` test runs (https://github.com/purescript/spago/blob/master/src/Spago/Build.hs#L288) because it basically calls `sha256 $ "output/Main/index.js <> "Nothing" <> "[ 'effect', 'console', 'prelude']" <> sameOpts`. Each of these values are the same between the `0.15.0` and `0.14.0` runs. This means someone who calls `spago script` on ` 0.15.0` compiler and then a similar script on `0.14.0` will get this same error.

There's two ways this could be fixed:
- include a boolean that indicates whether the PS compiler version used is ES or CJS modules
- include the compiler's major version in the `stableHash`.

I think I would favor the second.

**Comment by JordanMartinez** [comment]
The bower test was failing because I was checking out `v6.0.0` on both `0.15.0` and `0.14.0`. I should be checking out the `v5.0.0` tag on `0.14.0`:
```
2022-05-10 04:20:35.131356: [warn] 

Spago encountered some errors while trying to migrate your Bower config.
A Spago config has been generated but it's recommended that you apply the suggestions here

These packages are in the set, but did not match the Bower range. You can try to install them with `spago install some-package-name`
* "assert" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "console" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "control" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "effect" has version v3.0.0, but range is >=4.0.0 <5.0.0
* "invariant" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "maybe" has version v5.0.0, but range is >=6.0.0 <7.0.0
* "prelude" has version v5.0.0, but range is >=6.0.0 <7.0.0
```
Latest commit fixes that, so CI should now build.

Lastly, in [1114207](https://github.com/purescript/spago/pull/879/commits/11142077affd4b1d26d7f7cd41dd6fad743fa230), I always print out stdout/stderr if a command we expect to succeed doesn't. This should help us debug CI errors in the future as we won't need to push a commit enabling this printing when a test fails.

**Comment by JordanMartinez** [comment]
CI builds! :tada:

File: src/Spago/Build.hs [hunk 1]
```diff
@@ -278,14 +278,19 @@ script modulePath tag packageDeps opts = do
   absoluteModulePath <- fmap Text.pack (makeAbsolute (Text.unpack modulePath))
   currentDir <- Turtle.pwd
 
+  (Version.SemVer major minor _ _ _) <- Purs.pursVersion >>= \case
+    Left t -> die [ "Could not determine PureScript compiler version: " <> display t ]
+    Right v -> pure v
+
```
[the remark this answers is not in the document]
Comment (JordanMartinez) [reply src/Spago/Build.hs]: Just for documentation purposes...

We only need a String for the compiler version's part of the hash. `pursVersion` can fail for one of two reasons:
1. due to running `purs --version` successfully but failing to parse the version. In this situation, `spago script` should still succeed.
1. the command itself fails. In this situation, some later parts in `spago script` may fail in the future, but we should try to make progress by ignoring this issue.

In both cases of failure, we can get a String by using the text returned from the command.

As for the version case, there's a few factors to consider:
- while we're on the `0.x.x` series, the minor value represents the major value and the patch represents the minor value.
- running `spago script` 3 times, first on `0.14.4`, then on `0.14.5` (assuming `0.14.5` adds new syntax to the language like visible type applications), then on `0.14.4`, `spago script` should use 2 different folders, even if there are times where this isn't needed. This is a conservative approach that simplifies the logic here
- once PureScript starts making continuous deployments, an alpha release (e.g. `0.15.1-1`) may introduce things that are removed in a future alpha release (e.g. `0.15.1-2`). So, a proper solution would need to account for this possibility.

As a result, simply `show`ing the version to get a String for the hash is a conservative approach with simple logic for guaranteeing that this works correctly. The downside is that the script will create multiple folders with potentially the same stuff when the same script is run using different compiler versions. However, we think this tradeoff is worth it due to the infrequency of this situation occurring.

**Pull request merged**